theia-extensions/viewer-prototype/src/node/trace-server-service.ts

```diff
@@ -1,60 +1,88 @@
-            // That doesn't mean that the server has really started. On the frontend, the `TraceServerConnectionStatusService`
-            // will ping the port until it receives a response, which is the official measure of success.
 
-            timeouts.push(setTimeout(() => resolve('success'), 2000));
+    private server?: ChildProcess;
+    private port?: number;
 
-            // If the server exits or errors before it outputs, consider it a failure.
-            server.on('error', err => {
-                reject(err);
-            });
-            server.on('exit', code => {
-                reject(PortBusy(code));
+    async startTraceServer(path?: string, port?: number): Promise<string> {
+        if (path === undefined) {
+            throw new Error('no Trace Server path specified');
+        } else if (port === undefined) {
+            throw new Error('no Trace Server port specified');
+        } else if (this.isServerRunning(this.server)) {
+            if (this.port === port) {
+                return SUCCESS;
+            } else {
+                throw new Error('the Trace Server is already running on a different port');
+            }
+        }
+        const server = spawn(path, ['-vmargs', `-Dtraceserver.port=${port}`]);
+        if (server.pid === undefined) {
+            // When pid is undefined it usually means we're about to get an error.
+            return new Promise<never>((_, reject) => server.once('error', reject));
+        }
+        this.server = server;
+        this.port = port;
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        let timeout: any;
+        try {
+            await new Promise<void>((resolve, reject) => {
+                // If the server doesn't error or exit within 2 seconds, we consider that a success.
+                // That doesn't mean that the server has really started. On the frontend, the `TraceServerConnectionStatusService`
+                // will ping the port until it receives a response, which is the official measure of success.
+                timeout = setTimeout(resolve, 2000);
+                // If the server exits or errors before it outputs, consider it a failure.
+                server.once('exit', (code, _signal) => reject(PortBusy(code)));
+                server.once('error', reject);
             });
-            // If no response recieved from the trace server in 10 seconds, reject with an error (for internal use)
-            timeouts.push(setTimeout(() => reject('Waited 10 seconds but nothing happened and hence exiting'), 10000));
-
-        }).finally(() => {
+        } finally {
             server.removeAllListeners();
-            timeouts.forEach(timeout => clearTimeout(timeout));
-        });
-
+            // The process exited so we can clean our internal state.
+            server.once('exit', () => {
+                this.server = undefined;
+                this.port = undefined;
+            });
```

Comment:
Is this guaranteed to run even if the server has already exited (e.g. if we arrive here because we rejected the Promise due to an exited event)? I don't think it is, and since we have checks for whether the server is running that don't require us to remove these fields, and this listener isn't guaranteed to remove them, I don't this listener is necessary? Or, if we want this code to run when the server exits, it should be put in the exit handler defined in the Promise callback.

Reply:
Good catch, this was bad logic. This listener must only be added if the server process "successfully" starts.

Reply:
It looks like you've relocated this listener after the `finally`, but I think that leaves the same problem as before: if the `exit` event fires while we're `await`ing the Promise, then this listener will be applied too late.

Reply:
Touche ⚔️ . I missed the re`throw` in the `catch` block 👍.